src/EventStore.Core/Services/RedactionService.cs

```diff
@@ -0,0 +1,261 @@
+﻿using System;
+using System.IO;
+using System.Threading;
+using EventStore.Common.Utils;
+using EventStore.Core.Bus;
+using EventStore.Core.Data.Redaction;
+using EventStore.Core.Exceptions;
+using EventStore.Core.Messages;
+using EventStore.Core.Messaging;
+using EventStore.Core.Services.Storage.ReaderIndex;
+using EventStore.Core.Synchronization;
+using EventStore.Core.TransactionLog.Chunks;
+using EventStore.Core.TransactionLog.Chunks.TFChunk;
+using Serilog;
+
+namespace EventStore.Core.Services {
+	public class RedactionService<TStreamId> :
+		IHandle<RedactionMessage.GetEventPosition>,
+		IHandle<RedactionMessage.SwitchChunkLock>,
+		IHandle<RedactionMessage.SwitchChunk>,
+		IHandle<RedactionMessage.SwitchChunkUnlock>,
+		IHandle<SystemMessage.BecomeShuttingDown> {
+
+		private readonly IQueuedHandler _queuedHandler;
+		private readonly TFChunkDb _db;
+		private readonly IReadIndex<TStreamId> _readIndex;
+		private readonly SemaphoreSlimLock _switchChunksLock;
+
+		private const string NewChunkFileExtension = ".tmp";
+
+		public RedactionService(
+			IQueuedHandler queuedHandler,
+			TFChunkDb db,
+			IReadIndex<TStreamId> readIndex,
+			SemaphoreSlimLock switchChunksLock) {
+			Ensure.NotNull(queuedHandler, nameof(queuedHandler));
+			Ensure.NotNull(db, nameof(db));
+			Ensure.NotNull(readIndex, nameof(readIndex));
+			Ensure.NotNull(switchChunksLock, nameof(switchChunksLock));
+
+			_queuedHandler = queuedHandler;
+			_db = db;
+			_readIndex = readIndex;
+			_switchChunksLock = switchChunksLock;
+		}
+
+		public void Handle(RedactionMessage.GetEventPosition message) {
+			try {
+				GetEventPosition(message.EventStreamId, message.EventNumber, message.Envelope);
+			} catch (Exception ex) {
+				Log.Error(ex, "An error has occurred when getting position for stream: {stream}, event number: {eventNumber}.",
+					message.EventStreamId, message.EventNumber);
+				message.Envelope.ReplyWith(
+					new RedactionMessage.GetEventPositionCompleted(GetEventPositionResult.UnexpectedError, Array.Empty<EventPosition>()));
+			}
+		}
+
+		private void GetEventPosition(string streamName, long eventNumber, IEnvelope envelope) {
+			var streamId = _readIndex.GetStreamId(streamName);
+			var result = _readIndex.ReadEventInfo_KeepDuplicates(streamId, eventNumber);
+
+			var eventPositions = new EventPosition[result.EventInfos.Length];
+
+			for (int i = 0; i < result.EventInfos.Length; i++) {
+				var eventInfo = result.EventInfos[i];
+				var logPos = eventInfo.LogPosition;
+				var chunk = _db.Manager.GetChunkFor(logPos);
+				var localPosition = chunk.ChunkHeader.GetLocalLogPosition(logPos);
```

Comment:
not certain, but do we need to increase the maxreadercount for this

Reply:
well spotted, fixed :+1:

we actually need it for _readIndex.ReadEventInfo_KeepDuplicates() and chunk.GetActualRawPosition(), so i've added 1 to GetPTableMaxReaderCount()